packages/salesforcedx-apex-replay-debugger/src/core/heapDumpService.ts

```diff
@@ -136,49 +127,123 @@ export class HeapDumpService {
+        variableTypes.get(variableEntry.keyDisplayValue)
+      );
+    }
+  }
+
+  public updateContainer(
     varContainer: ApexVariableContainer,
-    heapDumpExtentValue: HeadpDumpExtentValue,
-    type?: string
+    heapDumpExtentValue: HeapDumpExtentValue,
+    extentToRevisit: Map<string, HeapDumpExtentValue>,
+    referenceToRevisit: Map<string, ApexVariableContainer>,
+    variableTypes: Map<string, string>
   ): void {
     if (heapDumpExtentValue.value) {
-      this.updateVariableContainerWithExtentValue(
+      this.updateContainerWithExtentValue(
         varContainer,
         heapDumpExtentValue.value,
-        type
+        variableTypes.get(varContainer.name)
       );
     } else if (heapDumpExtentValue.entry) {
-      varContainer.variables.forEach((value, key) => {
-        for (const variableEntry of heapDumpExtentValue.entry!) {
-          const valueAsApexVar = value as ApexVariableContainer;
-          this.updateVariableContainerWithExtentValueOrEntry(
+      this.updateContainerChildrenWithEntries(
+        varContainer,
+        heapDumpExtentValue.entry,
+        extentToRevisit,
+        referenceToRevisit,
+        variableTypes
+      );
+    }
+  }
+
+  public updateContainerChildrenWithEntries(
+    varContainer: ApexVariableContainer,
+    extentValueEntries: HeapDumpExtentValueEntry[],
+    extentToRevisit: Map<string, HeapDumpExtentValue>,
+    referenceToRevisit: Map<string, ApexVariableContainer>,
+    variableTypes: Map<string, string>
+  ): void {
+    for (const extentValueEntry of extentValueEntries) {
+      let foundMatchingApexVariable = false;
+      for (const entry of Array.from(varContainer.variables.entries())) {
+        const valueAsApexVar = entry[1] as ApexVariableContainer;
+        if (this.isContainerForExtentEntry(valueAsApexVar, extentValueEntry)) {
+          this.updateContainerOrCache(
             valueAsApexVar,
-            variableEntry.value.value,
-            variableEntry.value.entry
+            extentValueEntry,
+            extentToRevisit,
+            referenceToRevisit,
+            variableTypes
           );
+          foundMatchingApexVariable = true;
+          break;
         }
-      });
+      }
+      if (!foundMatchingApexVariable) {
+        const variableType =
+          variableTypes.get(extentValueEntry.keyDisplayValue) || '';
+        varContainer.variables.set(
+          extentValueEntry.keyDisplayValue,
+          new ApexVariableContainer(
+            extentValueEntry.keyDisplayValue,
+            extentValueEntry.value.value,
+            variableType
+          )
+        );
+      }
     }
   }
 
-  public updateVariableContainerWithExtentValueOrEntry(
+  public updateContainerWithExtentValueOrEntry(
     varContainer: ApexVariableContainer,
     extentValue: any,
-    extentEntry: HeadpDumpExtentValueEntry[] | undefined
+    extentEntry: HeapDumpExtentValueEntry[] | undefined,
+    type?: string
   ): void {
     if (extentValue) {
```

Comment:
Same issue where if the value happens to be a boolean set to false. We should be checking whether or not typeof extentValue !== 'undefined'